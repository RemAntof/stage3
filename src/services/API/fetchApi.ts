import {
  BASE_URL,
  DEFAULT_PAGE,
  DEFAULT_PAGESIZE,
} from '@constants/apiEndpoints';
import ApiResponse from '@interfaces/apiResponse';
import formUrlEncoded from 'form-urlencoded';

const fetchApi = async (
  name: string | undefined = undefined,
  activePage: number = DEFAULT_PAGE
): Promise<ApiResponse> => {
  const url = name
    ? `${BASE_URL}?pageNumber=${activePage}&pageSize=${DEFAULT_PAGESIZE}&name=${name}`
    : `${BASE_URL}?pageNumber=${activePage}&pageSize=${DEFAULT_PAGESIZE}`;
  const method = name ? 'POST' : 'GET';
  const body = name ? formUrlEncoded({ name: name }) : null;
  const response = await fetch(url, {
    method: method,
    headers: {
      accept: 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: body,
  });

  if (!response.ok) {
    throw new Error(`Error: ${response.statusText}`);
  }

  const data = await response.json();
  console.log(data);
  return data;
};
export default fetchApi;
