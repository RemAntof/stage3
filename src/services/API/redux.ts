import {
  BASE_URL,
  DEFAULT_PAGESIZE,
} from '@constants/apiEndpoints';
import ApiResponse from '@interfaces/apiResponse';
import {
  createApi,
  fetchBaseQuery,
} from '@reduxjs/toolkit/query/react';
import formUrlEncoded from 'form-urlencoded';

const api = createApi({
  baseQuery: fetchBaseQuery({
    baseUrl: BASE_URL,
    headers: {
      accept: 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  }),
  endpoints: (build) => ({
    animal: build.query<
      ApiResponse,
      { activePage: number; name: string }
    >({
      query({ activePage, name }) {
        return {
          url: '/search',
          params: {
            pageNumber: activePage,
            pageSize: DEFAULT_PAGESIZE,
          },
          method: name ? 'POST' : 'GET',
          body: name
            ? formUrlEncoded({ name: name })
            : undefined,
        };
      },
    }),
    animalDetails: build.query<
      ApiResponse,
      { name: string }
    >({
      query({ name }) {
        return {
          url: '/search',
          params: {
            name: name,
            pageNumber: 0,
            pageSize: DEFAULT_PAGESIZE,
          },
          method: name ? 'POST' : 'GET',
          body: name
            ? formUrlEncoded({ name: name })
            : undefined,
        };
      },
    }),
  }),
});

export default api;
export const { useAnimalQuery, useAnimalDetailsQuery } =
  api;
