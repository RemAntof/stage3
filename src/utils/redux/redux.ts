import {
  configureStore,
  createSlice,
  PayloadAction,
} from '@reduxjs/toolkit';

export interface FormDataState {
  name: string;
  email: string;
  age: number | null;
  gender: string;
  country: string;
  picture: string;
  terms: boolean;
  password: string;
}

const initialState: FormDataState = {
  name: '',
  email: '',
  age: null,
  gender: '',
  country: '',
  picture: '',
  terms: false,
  password: '',
};

const countriesSlice = createSlice({
  name: 'countries',
  initialState: [
    'USA',
    'Canada',
    'Mexico',
    'Germany',
    'Australia',
    'Poland',
    'Belarus',
  ],
  reducers: {},
});

const uncontrolledFormDataSlice = createSlice({
  name: 'formData',
  initialState,
  reducers: {
    saveUncontrolledFormData: (
      state,
      action: PayloadAction<FormDataState>
    ) => {
      return { ...state, ...action.payload };
    },
  },
});

const controlledFormDataSlice = createSlice({
  name: 'formData',
  initialState,
  reducers: {
    saveControlledFormData: (
      state,
      action: PayloadAction<FormDataState>
    ) => {
      return { ...state, ...action.payload };
    },
  },
});

export const { saveUncontrolledFormData } =
  uncontrolledFormDataSlice.actions;
export const { saveControlledFormData } =
  controlledFormDataSlice.actions;

const store = configureStore({
  reducer: {
    countries: countriesSlice.reducer,
    uncontrolledFormData: uncontrolledFormDataSlice.reducer,
    controlledFormData: controlledFormDataSlice.reducer,
  },
});

export default store;
