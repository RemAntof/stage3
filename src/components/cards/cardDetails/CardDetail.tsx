import React, { useRef } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import Loader from '@components/loader/loader';
import Card from '../card/card';

import ErrorPage from '@views/errorView/errorView';
import { useAnimalDetailsQuery } from '@services/API/redux';

const CardDetail: React.FC = () => {
  const { animalName } = useParams<{
    animalName: string;
  }>();
  const cardDetailRef = useRef<HTMLDivElement>(null);
  const navigate = useNavigate();
  const { isLoading, isError, data } =
    useAnimalDetailsQuery({ name: animalName });

  const navigateBack = () => {
    navigate(-1);
  };

  if (isLoading) return <Loader />;

  if (isError) return <ErrorPage />;
  const animal = data.animals[0];

  return (
    <div ref={cardDetailRef}>
      <button onClick={navigateBack}>Go Back</button>
      {animal ? (
        <Card animal={animal} />
      ) : (
        <div>Animal not found</div>
      )}
    </div>
  );
};

export default CardDetail;
