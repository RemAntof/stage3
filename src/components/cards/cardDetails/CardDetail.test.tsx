import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import CardDetail from './CardDetail';

import { useAnimalDetailsQuery } from '@services/API/redux';

jest.mock('@services/API/redux', () => ({
  useAnimalDetailsQuery: jest.fn(),
}));

jest.mock('@components/loader/loader', () => () => (
  <div>Loading...</div>
));
jest.mock('@views/errorView/errorView', () => () => (
  <div>Error occurred</div>
));
jest.mock('../card/card', () => ({ animal }) => (
  <div>Card for {animal.name}</div>
));

describe('CardDetail Snapshot Tests', () => {
  test('renders loading state correctly', () => {
    (useAnimalDetailsQuery as jest.Mock).mockReturnValue({
      isLoading: true,
      isError: false,
      data: null,
    });

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/animals/lion']}>
        <CardDetail />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders error state correctly', () => {
    (useAnimalDetailsQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      isError: true,
      data: null,
    });

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/animals/lion']}>
        <CardDetail />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders data state correctly', () => {
    (useAnimalDetailsQuery as jest.Mock).mockReturnValue({
      isLoading: false,
      isError: false,
      data: {
        animals: [
          {
            name: 'Lion',
            description: 'King of the jungle',
          },
        ],
      },
    });

    const { asFragment } = render(
      <MemoryRouter initialEntries={['/animals/lion']}>
        <CardDetail />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
