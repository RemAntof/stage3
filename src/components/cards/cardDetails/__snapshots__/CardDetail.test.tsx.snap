// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDetail Snapshot Tests renders data state correctly 1`] = `
<DocumentFragment>
  <div>
    <button>
      Go Back
    </button>
    <div>
      Card for Lion
    </div>
  </div>
</DocumentFragment>
`;

exports[`CardDetail Snapshot Tests renders error state correctly 1`] = `
<DocumentFragment>
  <div>
    Error occurred
  </div>
</DocumentFragment>
`;

exports[`CardDetail Snapshot Tests renders loading state correctly 1`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;

exports[`CardDetail component navigates back when Go Back button is clicked 1`] = `
<div>
  <div>
    <button>
      Go Back
    </button>
    <div
      class="cardBox cardBox light"
    >
      <div
        class="cardName cardName light"
      >
        Test Animal
      </div>
      <div
        class="cardText"
      >
        🌍Earth Animal:
         
        Yes
      </div>
      <div
        class="cardText"
      >
        🐛Earth Insect:
         
        NO
      </div>
      <div
        class="cardText"
      >
        🐦Avian: 
        NO
      </div>
      <div
        class="cardText"
      >
        🐕‍🦺Canine: 
        Yes
      </div>
      <div
        class="cardText"
      >
        🐈Feline: 
        NO
      </div>
    </div>
  </div>
</div>
`;

exports[`CardDetail component renders correctly when animal is fetched 1`] = `
<div>
  <div>
    <button>
      Go Back
    </button>
    <div
      class="cardBox"
    >
      <div
        class="cardName"
      >
        Test Animal
      </div>
      <div
        class="cardText"
      >
        🌍Earth Animal:
         
        Yes
      </div>
      <div
        class="cardText"
      >
        🐛Earth Insect:
         
        NO
      </div>
      <div
        class="cardText"
      >
        🐦Avian: 
        NO
      </div>
      <div
        class="cardText"
      >
        🐕‍🦺Canine: 
        Yes
      </div>
      <div
        class="cardText"
      >
        🐈Feline: 
        NO
      </div>
    </div>
  </div>
</div>
`;

exports[`CardDetail component renders error message when fetch fails 1`] = `
<div>
  <div>
    Error: 
    Fetch failed
  </div>
</div>
`;
